From 78b593b28f875582360afe47f73372b0f70af90f Mon Sep 17 00:00:00 2001
From: Mike Fulton <mikefultonpersonal@gmail.com>
Date: Fri, 17 Jun 2022 11:09:29 -0700
Subject: [PATCH 17/40] 016_EC_range

---
 crypto/ec/ec_lib.c | 18 ++++++++++++++++++
 1 file changed, 18 insertions(+)

diff --git a/crypto/ec/ec_lib.c b/crypto/ec/ec_lib.c
index aff1497..0d63e0d 100644
--- a/crypto/ec/ec_lib.c
+++ b/crypto/ec/ec_lib.c
@@ -804,6 +804,15 @@ int EC_POINT_set_affine_coordinates_GFp(const EC_GROUP *group,
                                         EC_POINT *point, const BIGNUM *x,
                                         const BIGNUM *y, BN_CTX *ctx)
 {
+    BIGNUM *p = 0;
+    p = BN_new();
+    EC_GROUP_get_curve_GFp(group,p,NULL,NULL,ctx);
+    if((BN_cmp(p,x) < 0) ||  (BN_cmp(p,y) < 0 ) ) {
+      ECerr(EC_F_EC_POINT_SET_AFFINE_COORDINATES_GFP,EC_R_POINT_IS_NOT_ON_CURVE);
+      BN_clear_free(p);
+      return 0;
+    }
+    BN_clear_free(p);
     return EC_POINT_set_affine_coordinates(group, point, x, y, ctx);
 }
 
@@ -812,6 +821,15 @@ int EC_POINT_set_affine_coordinates_GF2m(const EC_GROUP *group,
                                          EC_POINT *point, const BIGNUM *x,
                                          const BIGNUM *y, BN_CTX *ctx)
 {
+    BIGNUM *p = 0;
+    p = BN_new();
+    EC_GROUP_get_curve_GF2m(group,p,NULL,NULL,ctx);
+    if((BN_cmp(p,x) < 0) ||  (BN_cmp(p,y) < 0 ) ) {
+      ECerr(EC_F_EC_POINT_SET_AFFINE_COORDINATES_GF2M,EC_R_POINT_IS_NOT_ON_CURVE);
+      BN_clear_free(p);
+      return 0;
+    }
+    BN_clear_free(p);
     return EC_POINT_set_affine_coordinates(group, point, x, y, ctx);
 }
 # endif
-- 
2.36.1

