From ec7d2ca31aa9c7f8ca87557f93f965948b8edae2 Mon Sep 17 00:00:00 2001
From: Mike Fulton <mikefultonpersonal@gmail.com>
Date: Fri, 17 Jun 2022 11:09:32 -0700
Subject: [PATCH 34/40] 038_zosa

---
 Configurations/10-main.conf | 24 ++++++++++++++++++++++++
 1 file changed, 24 insertions(+)

diff --git a/Configurations/10-main.conf b/Configurations/10-main.conf
index e059470..d0a6643 100644
--- a/Configurations/10-main.conf
+++ b/Configurations/10-main.conf
@@ -1209,6 +1209,7 @@ my %targets = (
         thread_scheme    => "(unknown)",
     },
 # IBM zOS, 64 and 31 bit
+# 64 bit EBCDIC
      "OS390-Unix" => {
          inherit_from     => [ "BASE_unix", asm("zos_asm") ],
          cc               => "./tools/c99.sh",
@@ -1219,6 +1220,18 @@ my %targets = (
          thread_scheme    => "(unknown)",
 	 perlasm_scheme   => "zOS64",
      },
+# 64 bit ASCII
+     "OS390-ASCII" => {
+        inherit_from     => [ "BASE_unix", asm("zos_asm") ],
+        cc               => "./tools/c99.sh",
+        cflags           => "-O -Wc,dll,XPLINK,exportall,hgpr,lp64,ASCII  -Wa,'GOFF,SYSPARM(USE_XPLINK)' -qlongname -qlanglvl=extc99 -DB_ENDIAN -DNO_SYS_PARAM_H -D_ALL_SOURCE -D_OPEN_THREADS=2 -D_POSIX_SOURCE  -D_OPEN_MSGQ_EXT",
+        module_ldflags   => "-Wl,XPLINK,LP64",
+        shared_ldflags   => "-Wl,dll,XPLINK,LP64",
+        bn_ops           => "SIXTY_FOUR_BIT_LONG RC4_CHAR",
+        thread_scheme    => "(unknown)",
+        perlasm_scheme   => "zOS64",
+     },
+# 32 bit EBCDIC
      "OS390-Unix31" => {
          inherit_from     => [ "BASE_unix", asm("zos32_asm") ],
          cc               => "./tools/c99.sh",
@@ -1229,6 +1242,17 @@ my %targets = (
          thread_scheme    => "(unknown)",
          perlasm_scheme   => "zOS31",
     },
+# 32 bit ASCII
+     "OS390-ASCII31" => {
+         inherit_from     => [ "BASE_unix", asm("zos32_asm") ],
+         cc               => "./tools/c99.sh",
+         cflags           => "-O -Wc,'ARCH(9),dll,XPLINK,exportall,hgpr,ASCII'  -Wa,'GOFF,SYSPARM(USE_XPLINK)' -qlongname -qlanglvl=extc99 -DB_ENDIAN -DNO_SYS_PARAM_H -D_ALL_SOURCE -D_OPEN_THREADS=2 -D_POSIX_SOURCE  -D_OPEN_MSGQ_EXT",
+         module_ldflags   => "-Wl,XPLINK",
+         shared_ldflags   => "-Wl,dll,XPLINK",
+         bn_ops           => "SIXTY_FOUR_BIT RC4_CHAR",
+         thread_scheme    => "(unknown)",
+         perlasm_scheme   => "zOS31",
+    },
 #### Visual C targets
 #
 # Win64 targets, WIN64I denotes IA-64/Itanium and WIN64A - AMD64
-- 
2.36.1

