diff --git a/crypto/bn/bn_ctx.c b/crypto/bn/bn_ctx.c
index 042cb24..b6a3611 100644
--- a/crypto/bn/bn_ctx.c
+++ b/crypto/bn/bn_ctx.c
@@ -277,7 +277,10 @@ static int BN_STACK_push(BN_STACK *st, unsigned int idx)
 
 static unsigned int BN_STACK_pop(BN_STACK *st)
 {
-    return st->indexes[--(st->depth)];
+    if(st->depth) {
+        return st->indexes[--(st->depth)];
+    }
+    return 0;
 }
 
 /***********/
