diff --git a/crypto/bn/bn_lib.c b/crypto/bn/bn_lib.c
index eb4a318..e0102d3 100644
--- a/crypto/bn/bn_lib.c
+++ b/crypto/bn/bn_lib.c
@@ -223,12 +223,7 @@ void BN_clear_free(BIGNUM *a)
 
 void BN_free(BIGNUM *a)
 {
-    if (a == NULL)
-        return;
-    if (!BN_get_flags(a, BN_FLG_STATIC_DATA))
-        bn_free_d(a, 0);
-    if (a->flags & BN_FLG_MALLOCED)
-        OPENSSL_free(a);
+    BN_clear_free(a);
 }
 
 void bn_init(BIGNUM *a)
