diff --git a/crypto/bio/b_addr.c b/crypto/bio/b_addr.c
index 0af7a33..c71a7bf 100644
--- a/crypto/bio/b_addr.c
+++ b/crypto/bio/b_addr.c
@@ -22,6 +22,14 @@
 #include <openssl/buffer.h>
 #include "internal/thread_once.h"
 
+#ifndef NI_MAXSERV
+#define NI_MAXSERV 32
+#endif
+
+#ifndef NI_MAXHOST
+#define NI_MAXHOST 1025
+#endif 
+
 CRYPTO_RWLOCK *bio_lookup_lock;
 static CRYPTO_ONCE bio_lookup_init = CRYPTO_ONCE_STATIC_INIT;
 
