--- openssl-4.0.0/test/bio_dgram_test.c
+++ openssl-4.0.0/test/bio_dgram_test.c
@@ -166,6 +166,7 @@
     fd1 = BIO_socket(af, SOCK_DGRAM, IPPROTO_UDP, 0);
     if (fd1 < 0) {
 #ifdef EAFNOSUPPORT
+        TEST_info("BIO_socket failed with error %d (EAFNOSUPPORT is %d)", get_last_socket_error(), EAFNOSUPPORT);
         if (get_last_socket_error() == EAFNOSUPPORT) {
             testresult = TEST_skip("BIO_socket() failed - Address family not supported");
             goto err;
