diff --git a/util/perl/OpenSSL/config.pm b/util/perl/OpenSSL/config.pm
index 3748788..5c2714a 100755
--- a/util/perl/OpenSSL/config.pm
+++ b/util/perl/OpenSSL/config.pm
@@ -670,6 +670,11 @@ EOF
             return { target => "linux-x86_64" };
         }
       ],
+      [ 's390x-.*-zos',
+        sub {
+            return target => "zos-64";
+        }
+      ],
       [ '.*86-.*-linux2',
         sub {
             # On machines where the compiler understands -m32, prefer a
