diff --git a/test/recipes/90-test_tls13secrets.t b/test/recipes/90-test_tls13secrets.t
index 72bb60f..9a93196 100644
--- a/test/recipes/90-test_tls13secrets.t
+++ b/test/recipes/90-test_tls13secrets.t
@@ -12,6 +12,8 @@ use OpenSSL::Test::Utils;
 my $test_name = "test_tls13secrets";
 setup($test_name);
 
+plan skip_all => "Test is disabled on z/OS" if config('target') =~ m|^zos|;
+
 plan skip_all => "$test_name is not supported in this build"
     if disabled("tls1_3")
        || disabled("shared")
