diff --git i/crypto/rsa/rsa_ossl.c w/crypto/rsa/rsa_ossl.c
index 1e56c6d..11dc29f 100644
--- i/crypto/rsa/rsa_ossl.c
+++ w/crypto/rsa/rsa_ossl.c
@@ -229,6 +229,8 @@ static uintptr_t get_unique_thread_id(void)
         CRYPTO_THREAD_set_local_ex(CRYPTO_THREAD_LOCAL_TANDEM_ID_KEY, NULL, (void *)thread_id);
     }
     return thread_id;
+#elif defined(__MVS__)
+    return (uintptr_t)CRYPTO_THREAD_get_current_id().__;
 #else
     return (uintptr_t)CRYPTO_THREAD_get_current_id();
 #endif
