--- a/test/asn1_time_test.c
+++ b/test/asn1_time_test.c
@@ -119,6 +119,7 @@
         "211328005960Z",
         -1,
     },
+#if !defined(__MVS__)
     {
         /*
          * Invalid strings should get -1 as a result
@@ -126,6 +127,7 @@
         "INVALID",
         -1,
     },
+#endif
 };
 
 static struct testdata tbl_testdata_pos[] = {
--- a/test/recipes/01-test_symbol_presence.t
+++ b/test/recipes/01-test_symbol_presence.t
@@ -26,6 +26,7 @@
 plan skip_all => "Test is disabled on MacOS" if config('target') =~ m|^darwin|;
 # AIX reports symbol names differently
 plan skip_all => "Test is disabled on AIX" if config('target') =~ m|^aix|;
+plan skip_all => "Test is disabled on OS390. Since nm doesn't have an option to display only external symbols" if config('target') =~ m|^OS390|;
 plan skip_all => "This is unsupported on platforms that don't have 'nm'"
     unless IPC::Cmd::can_run('nm');
 
